utils/bfd.ml

```diff
@@ -0,0 +1,540 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                        Nicolas Ojeda Bar, LexiFi                       *)
+(*                                                                        *)
+(*   Copyright 2020 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+type error =
+  | Truncated_file
+  | Unrecognized of string
+  | Unsupported of string * int
+
+exception Error of error
+
+let name_at ?max buf start =
+  assert (start >= 0 && start <= Bytes.length buf);
+  let rec loop pos =
+    if pos >= Bytes.length buf ||
+       Bytes.get buf pos = '\000' ||
+       (match max with None -> false | Some n -> pos - start >= n)
+    then
+      Bytes.sub_string buf start (pos - start)
+    else
+      loop (succ pos)
+  in
+  loop start
+
+let array_find_map f a =
+  let rec loop i =
+    if i >= Array.length a then None
+    else begin
+      match f a.(i) with
+      | None -> loop (succ i)
+      | Some _ as r -> r
+    end
+  in
+  loop 0
+
+let array_find f a =
+  array_find_map (fun x -> if f x then Some x else None) a
+
+let unsigned_of_int32 n =
+  assert (n >= 0l);
```

Comment:
Likewise, this should probably also `raise` explicitly, as its inputs come from an external file that might have negative offsets for some reason.

Reply:
Thanks, amended as suggested.